### PR TITLE
Fixed #23521 -- Support changing model bases in migrations

### PR DESCRIPTION
**Pull request by Ian-Foote**
Also document how to use the new `AlterModelBases` operation to refactor a concrete model inheritance into a stand-alone model.

https://code.djangoproject.com/ticket/23521

### CODE REVIEW

File: django/db/migrations/operations/models.py [hunk 1]
```diff
@@ -483,6 +483,28 @@ def reduce(self, operation, app_label=None):
...
+        state.reload_model(app_label, self.model_name)
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        pass
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        pass
+
+    def describe(self):
+        return "Update %s bases to %s" % (self.model_name, self.bases)
```
Comment (felixxm) [inline django/db/migrations/operations/models.py]: I would rename `self.model_name` -> `self.name` (like in `CreateModel`, `DeleteModel`, or `RenameModel`).

File: django/db/migrations/operations/models.py [hunk 2]
```diff
@@ -483,6 +483,28 @@ def reduce(self, operation, app_label=None):
         return super().reduce(operation, app_label=app_label)
 
 
+class AlterModelBases(ModelOperation):
+    reduce_to_sql = False
+    reversible = True
+
+    def __init__(self, model_name, bases):
+        self.model_name = model_name
+        self.bases = bases
```
Comment (felixxm) [inline django/db/migrations/operations/models.py]: Should we add sanity-check for duplicates in `bases`, `_check_for_duplicates()`? (e.g. [`CreateModel.__init__()`](https://github.com/django/django/blob/ef082ebb84f00e38af4e8880d04e8365c2766d34/django/db/migrations/operations/models.py#L55-L59)).

File: docs/howto/writing-migrations.txt [hunk 3]
```diff
@@ -326,3 +326,35 @@ If you want to change an unmanaged model (:attr:`managed=False
...
+
+Disconnecting a subclass from a concrete parent model
+=====================================================
+
+If you have a model ``Parent`` and a subclass ``Child`` and you want to make
+the ``Child`` model independent you need to write a custom migration. The key
+parts are ``AlterModelBases`` to tell django what the new model hierarchy is
+and ``AlterField`` to convert the implicit ``OneToOneField`` into an implicit
+``AutoField`` primary key. The ``RenameField` is useful to avoid needing to add
+the ``parent_ptr`` field explicitly to the updated ``Child`` model.
```
Comment (felixxm) [inline docs/howto/writing-migrations.txt]: I'm not sure if starting from an example is a good way to describe use cases :thinking: `AlterModelBases` should be used also in other cases e.g. renaming parent.

Maybe:
```
If you want to change parent models with
:ref:`multi-table inheritance <multi-table-inheritance>` you need to write a
custom migration e.g. to rename a parent model or make a child model
independent. ...

File: django/db/migrations/operations/models.py (hunk 2 above)
Comment (charettes) [reply django/db/migrations/operations/models.py]: `name` should be used to follow `ModelOperation.__init__` signature and `super().__init__(name)` should be called.

**Comment by Pomax** [comment]
This is a slightly aging PR, but I just ran into this while working on refactor work for our foundation.mozilla.org project, with no docs or blog posts or even SO questions that explained how to properly perform this task, so I posted https://stackoverflow.com/questions/61665607/renaming-a-django-superclass-model-and-updating-the-subclass-pointers-correctly/61723620, which garnered an answer that pointed to this PR: it would be *fantastic* if this PR could find its way into the codebase, making that SO question a matter of "This works out of the box as of 3.x.y - for older versions, the original answer was: ..."

**Comment by Ian-Foote** [comment]
I might get back to this sometime, but I'm quite busy with work at the moment, so if someone else wants to take it and get it into shape, that's ok by me.

The major work that still needs to be done is integrating with `makemigrations`, which I've not gotten around to digging into since I left this PR.

**Comment by Pomax** [comment]
@felixxm why did this get closed?

**Comment by felixxm** [comment]
Because @Ian-Foote don't have time to work on it. You can feel-free to continue his work.

**Comment by Pomax** [comment]
Given that the PR is almost done, leaving it open with an invitation to get it finished would certainly be preferable: even if I don't have time, I _do_ have money to maybe find someone pull this over the finishing line, since this is an important PR that would have saved us several days of being blocked. Sponsoring that work is perfectly acceptable.

**Comment by imomaliev** [comment]
@Pomax, Hi, it’s me the author of the answer of your question on SO. I would like to work on this. But probably will need some guidance from core dev team.

**Comment by Pomax** [comment]
@felixxm would you know whether someone from the core team would be willing to mentor this issue?

**Comment by felixxm** [comment]
Core team doesn't exist anymore, you can ask for a guidance on the developers mailing list or on forum.

> Given that the PR is almost done. 

The remaining part is not trivial. 

>... , leaving it open with an invitation to get it finished would certainly be preferable: 

Not really, leaving abandoned PRs open doesn't change anything. Ticket is marked as a ticket with patch that needs improvement, so anyone can finish it.

> Sponsoring that work is perfectly acceptable.

We don't allow donors to directly sponsor specific feature, see ["Do you accept donations from companies/corporations"](https://www.djangoproject.com/foundation/donate/).

**Comment by Pomax** [comment]
That is unfortunate. @imomaliev it sounds like you'll want to pick up this thread over on https://code.djangoproject.com/ticket/23521

**Pull request closed without merging**